### PR TITLE
[Torch] Fix bugs for `Torch::AtenOneHotOp`

### PR DESCRIPTION
**Pull request by angelz913**
This PR fixes the bugs for `Torch::AtenOneHotOp` by:

1) Using `Torch::kUnknownSize` as the default value for `numClasses` in
   the pattern matching stage in `DecomposeAtenOneHotOp`
2) Adding `AtenIntScalarOp` to the patterns in `TorchToArith`
3) Handling both `int` and `float` types for `off` and `on` values in `TorchOnnxToTorch` conversion

It also includes:

1) A new test in `TorchToArith/basic.mlir`, for `torch.aten.Int.Scalar`, and
2) A new test in `decompose-complex-ops.mlir`, for `torch.aten.one_hot`

**Dependencies**

This PR is dependent on #3334.

### CODE REVIEW

**Review by kuhar — COMMENTED** [review]
Are we able to write LIT tests for these?

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 1]
```diff
@@ -7064,10 +7064,7 @@ class DecomposeAtenOneHotOp : public OpRewritePattern<AtenOneHotOp> {
       return rewriter.notifyMatchFailure(
           op, "input tensor should have known sizes.");
     int64_t inputRank = inputType.getSizes().size();
-    int64_t numClasses;
-    if (!matchPattern(op.getNumClasses(), m_TorchConstantInt(&numClasses)))
```
Comment (kuhar) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:7068]: Should we try to constant fold but not emit the error when we fail?

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 2]
```diff
@@ -1665,12 +1665,12 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
         Value off = rewriter.create<Torch::AtenSelectIntOp>(loc, selectTy,
                                                             values, zero, zero);
         off = rewriter.create<Torch::AtenItemOp>(
-            loc, rewriter.getType<Torch::IntType>(), off);
+            loc, rewriter.getType<Torch::FloatType>(), off);
```
Comment (renxida) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:1668]: are we sure that it'll always be a float type?

File: lib/Conversion/TorchToArith/TorchToArith.cpp [hunk 3]
```diff
@@ -439,9 +439,10 @@ class ConvertTorchToArith
     target.addIllegalOp<Torch::ConstantIntOp>();
     patterns.add<ConvertTorchConstantIntOp>(typeConverter, context);
 
-    target.addIllegalOp<AtenIntBoolOp, AtenFloatScalarOp>();
+    target.addIllegalOp<AtenIntBoolOp, AtenFloatScalarOp, AtenIntScalarOp>();
```
Comment (renxida) [inline lib/Conversion/TorchToArith/TorchToArith.cpp:442]: Why this change? just curious.

**Review by renxida — APPROVED** [review]

File: lib/Conversion/TorchToArith/TorchToArith.cpp (hunk 3 above)
Comment (angelz913) [reply lib/Conversion/TorchToArith/TorchToArith.cpp:442]: I was following the examples of the `AtenIntBoolOp` and `AtenFloatScalarOp` (they were also marked as illegal op and a pattern was added for each).

**Review by renxida — APPROVED** [review]

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp (hunk 2 above)
Comment (angelz913) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp]: Now it handles both int and float types

File: test/Conversion/TorchToLinalg/onehot.mlir [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+// CHECK:               %[[EXT_0:.*]] = arith.extui %in : i1 to i64
+// CHECK:               linalg.yield %[[EXT_0]] : i64
+// CHECK:             } -> tensor<3x?xi64>
+// CHECK:             return %[[GENERIC_1]] : tensor<3x?xi64>
+func.func @torch.aten.one_hot(%arg0: !torch.vtensor<[3],si64>, %arg1: !torch.vtensor<[],f32>, %arg2: !torch.vtensor<[2],f32>) -> !torch.vtensor<[3,?],si64> {
+  %0 = torch.aten.item %arg1 : !torch.vtensor<[],f32> -> !torch.float
+  %1 = torch.aten.Int.Scalar %0 : !torch.float -> !torch.int
+  %2 = torch.aten.one_hot %arg0, %1 : !torch.vtensor<[3],si64>, !torch.int -> !torch.vtensor<[3,?],si64>
+  return %2 : !torch.vtensor<[3,?],si64>
+}
```
Comment (kuhar) [inline test/Conversion/TorchToLinalg/onehot.mlir:42]: nit: missing newline

File: test/Conversion/TorchToLinalg/onehot.mlir [hunk 5]
```diff
@@ -0,0 +1,42 @@
+// RUN: torch-mlir-opt --pass-pipeline="builtin.module(func.func(torch-decompose-complex-ops),torch-backend-to-linalg-on-tensors-backend-pipeline)" -verify-diagnostics < %s | FileCheck %s
+
+// -----
```
Comment (kuhar) [inline test/Conversion/TorchToLinalg/onehot.mlir:3]: You don't need this separator if you only have one test module. This separator is to allow for multiple modules in a single test, and this is enabled in the compiler using the `--split-input-file` flag.

File: test/Conversion/TorchToLinalg/onehot.mlir [hunk 6]
```diff
@@ -0,0 +1,42 @@
+// RUN: torch-mlir-opt --pass-pipeline="builtin.module(func.func(torch-decompose-complex-ops),torch-backend-to-linalg-on-tensors-backend-pipeline)" -verify-diagnostics < %s | FileCheck %s
```
Comment (qedawkins) [inline test/Conversion/TorchToLinalg/onehot.mlir:1]: I have not been involved with the ONNX work, but the test here looks like it's showing an integration test for changes in multiple places. I would expect to see a single test for the `AtenIntScalar` op lowering to arith added in this PR, and then separate tests for how the `onnx.one_hot` -> `torch.aten.one_hot` changed, and for how the `torch.aten.one_hot` decomposition changed, rather than testing all the way to linalg in one shot for this specific case.

There is a lit test file for testing changes to `DecomposeComplexOps` here: https://github.com/llvm/torch-mlir/blob/main/test/Dialect/Torch/decompose-complex-ops.mlir

And it looks to me that the `aten.scalar.int` to Arith test can go here: https://github.com/llvm/torch-mlir/blob/main/test/Conversion/TorchToArith/basic.mlir

The onnx test (if needed) can go here: https://github.com/llvm/torch-mlir/blob/main/test/Conversion/TorchOnnxToTorch/simple_ops_g_to_p.mlir

File: lib/Conversion/TorchToArith/TorchToArith.cpp (hunk 3 above)
Comment (qedawkins) [reply lib/Conversion/TorchToArith/TorchToArith.cpp:442]: Yeah torch-mlir does a partial conversion in each of its parts, so each lowering needs to mark the ops it's supposed to lower as illegal. It mainly helps report an error in the right place.

File: test/Conversion/TorchToLinalg/onehot.mlir (hunk 5 above)
[the remark this answers is not in the document]
Comment (qedawkins) [reply test/Conversion/TorchToLinalg/onehot.mlir:3]: (duplicate of Jakub's comment, closing)

**Review by kuhar — APPROVED** [review]
LGTM

**Comment by kuhar** [comment]
@qedawkins can you take another look and merge if everything's addressed?

**Review by qedawkins — APPROVED** [review]
LGTM, thanks!

**Comment by qedawkins** [comment]
Looks like the CI is failing for some reason. I can help merge after that's fixed.

**Comment by renxida** [comment]
automerge enabled. Fix the CI and we're good to go!
